src/package-resolver.js

```diff
@@ -545,37 +530,10 @@ export default class PackageResolver {
       this.resolveToResolution(req);
     }
 
-    for (const dep of deps) {
-      const name = normalizePattern(dep.pattern).name;
-      this.optimizeResolutions(name);
-    }
-
     activity.end();
     this.activity = null;
   }
 
-  // for a given package, see if a single manifest can satisfy all ranges
-  optimizeResolutions(name: string) {
-    const patterns: Array<string> = this.dedupePatterns(this.patternsByPackage[name] || []);
-
-    // don't optimize things that already have a lockfile entry:
-    // https://github.com/yarnpkg/yarn/issues/79
-    const collapsablePatterns = patterns.filter(pattern => {
-      const remote = this.patterns[pattern]._remote;
-      return !this.lockfile.getLocked(pattern) && (!remote || remote.type !== 'workspace');
-    });
-    if (collapsablePatterns.length < 2) {
-      return;
-    }
-
-    const availableVersions = this.getAllInfoForPatterns(collapsablePatterns).map(manifest => manifest.version);
-    const combinedRange = collapsablePatterns.map(pattern => normalizePattern(pattern).range).join(' ');
```

Comment:
Could #4547 also be fixed by removing this semver range munging (`patterns.join(' ')`), and testing each version against every pattern in turn? This seems to be basically what `semver.maxSatisfying` is doing, but would do the correct thing with ranges containing "||".

Reply:
Not really. My initial approach was to do a proper intersection and I have the code locally. The problem is `range` can be exotic too like `file:./a`.